### PR TITLE
Add INpcService definition

### PR DESCRIPTION
**Pull request by ivanbiljan**

### CODE REVIEW

**Comment by tylerjwatson** [comment]
I am good with this, thanks for your contribution!

If anyone has any problems with this patch let me know, I will merge it soon.

File: Orion/Npcs/NpcService.cs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+		/// <inheritdoc/>
+		public IEnumerable<INpc> FindNpcs(Predicate<INpc> predicate = null)
+		{
+			var npcs = new List<INpc>();
+			for (int i = 0; i < Terraria.Main.npc.Length; i++)
+			{
+				if (_npcs[i]?.WrappedNpc != Terraria.Main.npc[i])
+				{
+					_npcs[i] = new Npc(Terraria.Main.npc[i]);
```
Comment (Ijwu) [inline Orion/Npcs/NpcService.cs:109]: Why is this assignment happening? Is this just to sync up the `_npcs` array with the one in `Terraria.Main`?

**Comment by ivanbiljan** [comment]
Yes, it's just to keep everything synced. Other services have the same implementation:

```
        public IEnumerable<IItem> FindItems(Predicate<IItem> predicate = null)
        {
            var items = new List<IItem>();
            for (var i = 0; i < _items.Length; ++i)
            {
                if (_items[i]?.WrappedItem != Terraria.Main.item[i])
                {
                    _items[i] = new Item(Terraria.Main.item[i]);
                }
                items.Add(_items[i]);
            }
            return items.Where(i => i.WrappedItem.active && (predicate?.Invoke(i) ?? true));
        }
```

**Comment by Ijwu** [comment]
There's no other use of the `_npcs` array other than to sync it there. Do other services actually make use of their arrays?

**Comment by ivanbiljan** [comment]
Services like `PlayerService` do use it, e.g invoking the player joined event. After looking at it, `NpcService` should make use of its array when spawning an NPC. This will be fixed in the next commit.

**Comment by Ijwu** [comment]
Been informed by @MarioE 

`@ijwu: the use of the array is to cache `Npc` instances
so you don't `new` one every time`

Ok. That covers my only concern. I'm good for this merging after @ProfessorXZ's next commit. :+1:

File: Orion/Npcs/NpcService.cs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	/// <summary>
+	/// Manages <see cref="INpc"/> instances.
+	/// </summary>
+	[Service("Npc Service", Author = "Nyx Studios")]
+	public class NpcService : SharedService, INpcService
+	{
+		private readonly INpc[] _npcs;
+
+		/// <inheritdoc/>
+		public int BaseNpcSpawningLimit
```
Comment (MarioE) [inline Orion/Npcs/NpcService.cs:22]: Could you add tests for this and `BaseNpcSpawningRate`?

File: Orion.Tests/Npcs/NpcServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+		[TestCaseSource(nameof(NpcSpawnedTestCases))]
+		public void NpcSpawning_IsCorrect(NpcType type)
+		{
+			using (var orion = new Orion())
+			{
+				var npcService = orion.GetService<NpcService>();
+				var terrariaNpc = new Terraria.NPC();
+				var npc = new Npc(terrariaNpc);
+				var eventOccurred = false;
+				EventHandler<NpcSpawningEventArgs> handler = (sender, e) => eventOccurred = true;
```
Comment (MarioE) [inline Orion.Tests/Npcs/NpcServiceTests.cs:331]: This test could be strengthened by checking to see if the properties are populated correctly. Ditto for the above test,

File: Orion/Npcs/INpcService.cs [hunk 4]
```diff
@@ -37,7 +37,21 @@ public interface INpcService : IService
 		event EventHandler<NpcKilledEventArgs> NpcKilled;
 
 		/// <summary>
+<<<<<<< cc33da6b11c8d846a6d3e46ac7c60f9e027bfc6e
```
Comment (tylerjwatson) [inline Orion/Npcs/INpcService.cs]: You got merge crap in your NPC service definition, this won't compile.  Fix this and the build should work (Y)

**Comment by MarioE** [comment]
You are still dropping around twice the normal number of items every time an NPC dies, which is bad. Once in `DroppingLoot` and once in the actual `NPC.DropLoot` method. Your `DroppedLoot` won't work either if it finds the incorrect item because there are multiple items of that type in the world, too.

`_itemService` can also be `readonly`.

**Comment by Patrikkk-zz** [comment]
I think there is some misunderstanding here.
"Your DroppedLoot won't work either if it finds the incorrect item because there are multiple items of that type in the world, too."

He is not getting the existing item in the world. He's taking the type (item ID) what is picked and sent by terraria code when an NPC dies. ProfessorX simply creates an `Item` object which then gets passed to the event args along with the `NPC` object. 
it is only to make the life of plugin developers easier. X could just pass the item type through, and plugin devs would need to find item by ID them selves in their plugin.

I'm a bit out of the loop right now, especially with orion. But, how is he spawning items twice? o.0 don't see it.

**Comment by MarioE** [comment]
This gets existing items in the world: `var item = _itemService.FindItems(i => (int)i.Type == type).FirstOrDefault();`

EDIT: And nevermind about dropping items twice... that was `CreateItem` actually.

Oh, and also, maybe functionality should be added so that the `ref` arguments get modified.

**Comment by QuiCM** [comment]
Tagging in @NyxStudios/tshock for further commentary on this. Are we all good for a conflict resolution + merge?

**Comment by tylerjwatson** [comment]
le's keep the ball rolling, I say merge it and keep going

**Comment by Ijwu** [comment]
Manually merged. Please check my work @NyxStudios/tshock and let me know. 

Is there anything we need to keep going on @tylerjwatson or should we close this out?

**Comment by hakusaro** [comment]
@ijwu Please fix your git settings and claim your email on your account.
screen shot 2016-10-11 at 10 22 55 am

**Comment by Ijwu** [comment]
I have claimed the email address. Sorry, I put the merge through at work which required setting up a new git environment.

**Comment by hakusaro** [comment]
👍

**Pull request merged**